sympy/polys/series/ring.py

```diff
@@ -1,109 +1,1007 @@
+
+    def __repr__(self) -> str:
+        return (
+            f"Power Series Ring in {self.symbol} over {self.domain} of size {self.prec}"
+        )
+
+    def __eq__(self, other) -> bool:
+        return self.ring == other.ring
+
+    def is_element(self, element) -> TypeIs[PowerSeriesElement[Er]]:
+        """Check if element belongs to this ring."""
+        return isinstance(element, PowerSeriesElement) and element.ring == self
+
+    def order_term(self):
+        """
+        Return the order term of the power series.
+
+        Examples
+        ========
+        >>> from sympy import ZZ
+        >>> from sympy.polys.series import PowerSeriesRingRing
+        >>> R = PowerSeriesRingRing(ZZ, "x")
+        >>> R.order_term()
+        O(x**6)
+        >>> R10 = PowerSeriesRingRing(ZZ, "x", 10)
+        >>> R10.order_term()
+        O(x**10)
+
+        """
+        o = self.ring([], self.prec)
+        return self.from_element(o)
+
+    def from_expr(self, expr) -> PowerSeriesElement[Er]:
+        """
+        Convert an expression to a power series element.
+
+        Examples
+        ========
+
+        >>> from sympy import ZZ
+        >>> from sympy.polys.series import PowerSeriesRingRing
+        >>> R = PowerSeriesRingRing(ZZ, "x", 5)
+        >>> from sympy.abc import x
+        >>> s = R.from_expr(x + x**2 + x**4 + x**7); s
+        x + x**2 + x**4 + O(x**5)
+        >>> type(s)
+        <class 'sympy.polys.series.ring.PowerSeriesElement'>
+
+        """
+        poly = self._rebuild_expr(sympify(expr))
+        return self.ring_new(poly)
+
+    def _rebuild_expr(self, expr) -> TElement:
+        """Build the sympy expr into Lower Power Series format"""
+        prec = None
+        fring_prec: bool = False
+
+        if expr.has(Order):
+            var, p = expr.getO().expr.args
+
+            if var != self.symbol:
+                raise ValueError("Order contains different symbol than ring.")
+
+            if int(p) > self.prec:
+                prec = self.prec
+            else:
+                prec = int(p)
+
+            expr = expr.removeO()
+
+        max_prec = prec if prec else self.prec
+        coeffs = [self.domain.zero] * max_prec
+
+        for term in Add.make_args(expr):
+            coeff = self.domain.one
+            pow = 0
+
+            for factor in Mul.make_args(term):
+                if factor.is_Number:
+                    coeff = self.domain_new(factor)
+                else:
+                    base, exp = decompose_power(factor)
+                    if base == self.symbol:
+                        pow = int(exp)
+                    else:
+                        raise ValueError("Expr contains different symbol than ring.")
+
+            if pow < max_prec:
+                coeffs[pow] = coeff
+            else:
+                fring_prec = True
+
+        if fring_prec:
+            prec = self.prec
+
+        while coeffs and coeffs[-1] == self.domain.zero:
+            coeffs.pop()
```

Comment:
I think it would be better to collect a dict for coeffs rather than making a (potentially large) list and popping off the zeros after.